File: frontend/src/services/documentService.js
import api from './api'
import { FIELD_TYPES } from '../config/documentTypes'

function mapApiTypeToFieldType(apiType) {
  if (apiType.includes('String')) return FIELD_TYPES.STRING
  if (apiType.includes('Integer')) return FIELD_TYPES.INTEGER
  if (apiType.includes('Double')) return FIELD_TYPES.DOUBLE
  if (apiType.includes('DateTime')) return FIELD_TYPES.DATETIME
  return FIELD_TYPES.STRING
}

export async function fetchDocumentTypes() {
  const res = await api.get('/documents/document-types')
  return res.documentTypes.map((dt) => ({
    id: dt.name,
    labelKey: dt.name,
    country: dt.description.split(' - ')[0] || '',
    description: dt.description,
    processes: dt.processes,
  }))
}

export async function fetchActions(_documentTypeId, _process) {
  return []
}

export async function fetchDynamicFilters(documentTypeId) {
  const res = await api.get(`/documents/document-types/${documentTypeId}/filters`)
  return res.filters
    .filter((f) => f.isFilter)
    .map((f) => ({
      tag: f.name,
      labelKey: f.name,
      label: f.description,
      type: f.isRadioButton ? FIELD_TYPES.RADIO : mapApiTypeToFieldType(f.type),
      apiType: f.type,
      format: f.format,
      choices: f.choices,
    }))
}

export async function searchDocuments(filters) {
  const apiFilters = []
  if (filters.dynamicFilters) {
    for (const [tag, val] of Object.entries(filters.dynamicFilters)) {
      if (!val) continue
      if (typeof val === 'object' && val.from !== undefined) {
        if (val.from || val.to) {
          apiFilters.push({ name: tag, type: 'range', values: [val.from || '', val.to || ''] })
        }
      } else {
        apiFilters.push({ name: tag, type: 'string', value: String(val) })
      }
    }
  }

  const body = {
    documentType: filters.documentType,
    process: filters.process ? parseInt(filters.process, 10) : -1,
    situation: filters.situation ? parseInt(filters.situation, 10) : -1,
    historic: filters.historic || false,
    filters: apiFilters,
    pagination: { page: 1, pageSize: 50 },
    sort: { field: 'CreationDate', direction: 'desc' },
  }

  const res = await api.post('/documents/search', body)

  return {
    totalRecords: res.totalRecords,
    columns: res.columns.map((c) => ({ tag: c.name, labelKey: c.name, label: c.description })),
    rows: res.documents.map((doc) => {
      const row = {
        id: doc.documentId,
        statusDescription: doc.statusDescription,
        semaphore: doc.semaphore,
        createdAt: new Date().toISOString(),
      }
      if (doc.tags) {
        for (const [k, v] of Object.entries(doc.tags)) {
          row[k] = v
        }
      }
      return row
    }),
  }
}

export async function fetchDocumentDetail(documentId) {
  const doc = await api.get(`/documents/${documentId}`)
  return {
    id: doc.documentId,
    documentType: doc.documentType,
    process: doc.processType === 0 ? 'Outbound' : 'Inbound',
    statusDescription: doc.statusDescription,
    semaphore: doc.semaphore,
    createdAt: doc.creationDate,
    createdAtUTC: doc.creationDate,
    storePath: '/store/documents/' + doc.documentId,
    useDatabase: doc.historic === 0 ? 'Current' : 'History',
    owner: { companyId: doc.owner, searchCode: doc.ownerSearchCode || '' },
    receiver: { companyId: doc.receiver, searchCode: doc.receiverSearchCode || '' },
    tags: doc.tags || {},
  }
}

export async function fetchDocumentHistory(documentId) {
  const res = await api.get(`/documents/${documentId}/history`)
  return (res.history || []).map((h) => ({
    historyId: h.historyId,
    flowId: `FLOW-${h.flowId}`,
    creationDate: h.creationDate,
    statusDescription: h.statusDescription,
    semaphore: h.semaphore,
    processName: h.processName,
  }))
}

export async function fetchDocumentAttachments(documentId) {
  const res = await api.get(`/documents/${documentId}/attachments`)
  return (res.attachments || []).map((a) => ({
    id: a.attachmentId,
    name: a.attachmentName,
    type: a.type,
    creationDate: a.creationDate,
    downloadUrl: a.downloadUrl,
    contentUrl: a.contentUrl,
    canCopyContent: a.canCopyContent,
  }))
}

export async function fetchAttachmentContent(documentId, attachmentName) {
  try {
    const res = await api.get(`/documents/${documentId}/attachments/${attachmentName}/content`)
    const ct = res.contentType || ''
    let type = 'text'
    if (ct.includes('xml')) type = 'xml'
    else if (ct.includes('pdf')) type = 'pdf'
    return { type, content: res.content }
  } catch {
    return { type: 'pdf', content: null, message: 'Preview not available. Please download.' }
  }
}

export async function fetchDocumentErrors(documentId) {
  const res = await api.get(`/documents/${documentId}/errors`)
  return (res.errors || []).map((e) => ({
    id: e.errorId,
    creationDate: e.creationDate,
    method: e.method,
    message: e.message,
  }))
}

export async function fetchDocumentAudit(documentId) {
  const res = await api.get(`/documents/${documentId}/audit`)
  return (res.actions || []).map((a) => ({
    id: a.actionId,
    creationDate: a.creationDate,
    username: a.username,
    actionDescription: a.actionDescription,
  }))
}

export async function fetchDocumentFlags(documentId) {
  const res = await api.get(`/documents/${documentId}/flags`)
  return res.flags || []
}

export async function fetchDocumentMessages(documentId) {
  const res = await api.get(`/documents/${documentId}/messages`)
  return (res.messages || []).map((m) => ({
    id: m.messageId,
    creationDate: m.creationDate,
    messageType: m.messageType,
    code: m.code,
    description: m.description,
    customReturn: m.customReturn || '',
    sent: m.sent,
    error: m.error,
    processType: m.processType,
    tagVarchar1: m.tagVarchar1 || '',
    tagVarchar2: m.tagVarchar2 || '',
    tagVarchar3: m.tagVarchar3 || '',
  }))
}

export async function executeAction(documentId, _actionName, _formData) {
  return { success: true, message: `Action executed on document ${documentId}` }
}

export async function executeBatchAction(documentIds, actionName) {
  return { success: true, message: `Action "${actionName}" executed on ${documentIds.length} documents` }
}

export async function deleteDocument(documentId, justification) {
  return { success: true, message: `Document ${documentId} deleted. Reason: ${justification}` }
}

export async function sendToQueue(documentId, queueName) {
  return { success: true, message: `Document ${documentId} sent to queue "${queueName}"` }
}

export async function exportDocuments(_filters) {
  return { success: true, message: 'Export generated (mock - API endpoint pending)' }
}

File: api/src/Infrastructure/Services/DocumentService.cs
using Microsoft.EntityFrameworkCore;
using Sovos.Invoiceware.Application.DTOs.Documents;
using Sovos.Invoiceware.Application.Interfaces;
using Sovos.Invoiceware.Domain.Entities.Documents;
using Sovos.Invoiceware.Infrastructure.Data;

namespace Sovos.Invoiceware.Infrastructure.Services;

public class DocumentService : IDocumentService
{
    private readonly AppDbContext _db;
    public DocumentService(AppDbContext db) => _db = db;

    private static readonly List<DocumentTypeDto> DocTypes =
    [
        new("ide", "BR - Nota Fiscal Mercantil NF-e", [new(0, "Outbound"), new(1, "Inbound")]),
        new("CFDI", "MX - CFDI", [new(0, "Outbound"), new(1, "Inbound")]),
        new("DTE", "CL - Documento Tributario Electrónico", [new(0, "Outbound"), new(1, "Inbound")]),
        new("NFSe", "BR - Nota Fiscal de Serviço Eletrônica", [new(0, "Outbound")]),
        new("cteProc", "BR - Conhecimento de Transporte Eletrônico", [new(0, "Outbound"), new(1, "Inbound")]),
    ];

    private static readonly Dictionary<string, List<FilterDefinitionDto>> FilterDefs = new()
    {
        ["ide"] =
        [
            new("TagVarchar1", "CNPJ Emissor", "TF.Integration.Types.String", true, true, false, null, []),
            new("TagVarchar2", "CNPJ Receptor", "TF.Integration.Types.String", true, true, false, null, []),
            new("TagVarchar3", "Chave de Acesso", "TF.Integration.Types.String", true, true, false, null, []),
            new("TagInt1", "Série", "TF.Integration.Types.Integer", true, true, false, null, []),
            new("TagInt2", "Número NF", "TF.Integration.Types.Integer", true, true, false, null, []),
            new("TagFloat1", "Valor Total", "TF.Integration.Types.Double", true, true, false, "#,##0.00", []),
            new("TagDate1", "Data Emissão", "TF.Integration.Types.DateTime", true, true, false, null, []),
            new("TagInt5", "Status SEFAZ", "TF.Integration.Types.Integer", true, true, true, null,
                [new("-1", "Ignorar"), new("1", "Aprovado"), new("2", "Rejeitado"), new("0", "Não Contém")]),
        ],
        ["CFDI"] =
        [
            new("TagVarchar1", "RFC Emisor", "TF.Integration.Types.String", true, true, false, null, []),
            new("TagVarchar2", "RFC Receptor", "TF.Integration.Types.String", true, true, false, null, []),
            new("TagInt2", "Folio", "TF.Integration.Types.Integer", true, true, false, null, []),
            new("TagDate1", "Fecha Emisión", "TF.Integration.Types.DateTime", true, true, false, null, []),
        ],
    };

    public Task<List<DocumentTypeDto>> GetDocumentTypesAsync(CancellationToken ct = default)
        => Task.FromResult(DocTypes);

    public Task<DocumentTypeFiltersResponse> GetFiltersForTypeAsync(string documentType, CancellationToken ct = default)
    {
        var filters = FilterDefs.GetValueOrDefault(documentType, []);
        var columns = filters.Where(f => f.IsColumn).Select(f => new ColumnDefinitionDto(f.Name, f.Description, f.Type, f.Format)).ToList();
        return Task.FromResult(new DocumentTypeFiltersResponse(documentType, filters, columns));
    }

    public async Task<DocumentSearchResponse> SearchAsync(DocumentSearchRequest req, CancellationToken ct = default)
    {
        var page = req.Pagination?.Page ?? 1;
        var pageSize = Math.Min(req.Pagination?.PageSize ?? 25, 100);

        IQueryable<Document> query = _db.Set<Document>().AsNoTracking();

        query = query.Where(d => d.DocumentType == req.DocumentType);
        if (req.Process >= 0) query = query.Where(d => d.ProcessType == req.Process);
        if (req.Situation >= 0) query = query.Where(d => d.Semaphore == req.Situation);
        if (req.Historic) query = query.Where(d => d.Historic == 1);

        if (req.Filters != null)
        {
            foreach (var f in req.Filters)
            {
                query = ApplyFilter(query, f);
            }
        }

        var total = await query.CountAsync(ct);
        var dir = req.Sort?.Direction?.ToLower() == "asc";
        query = dir ? query.OrderBy(d => d.CreationDate) : query.OrderByDescending(d => d.CreationDate);
        var docs = await query.Skip((page - 1) * pageSize).Take(pageSize).ToListAsync(ct);

        var filterDefs = FilterDefs.GetValueOrDefault(req.DocumentType, []);
        var columns = filterDefs.Where(f => f.IsColumn).Select(f => new ColumnDefinitionDto(f.Name, f.Description, f.Type, f.Format)).ToList();

        var items = docs.Select(d => new DocumentListItemDto(
            d.DocumentId, d.StatusDescription, d.Semaphore, d.ProcessType, d.Historic,
            BuildTags(d, filterDefs)
        )).ToList();

        return new DocumentSearchResponse(total, page, pageSize, columns, items);
    }

    public async Task<DocumentDetailDto?> GetByIdAsync(string documentId, int historic = 0, CancellationToken ct = default)
    {
        var doc = await _db.Set<Document>().AsNoTracking().FirstOrDefaultAsync(d => d.DocumentId == documentId, ct);
        if (doc == null) return null;

        var filterDefs = FilterDefs.GetValueOrDefault(doc.DocumentType, []);
        var tags = BuildTags(doc, filterDefs);
        var tagDefs = filterDefs.Where(f => f.IsColumn).ToList();

        var actions = new List<ActionDto>
        {
            new(1, "Action::Reprint", "Re-imprimir", false, []),
            new(2, "Action::Cancel", "Cancelar", true, [new("xJustCanc", "Justificativa", "String", true)]),
        };

        var links = new Dictionary<string, string>
        {
            ["self"] = $"/api/v1/documents/{documentId}?historic={historic}",
            ["history"] = $"/api/v1/documents/{documentId}/history?historic={historic}",
            ["errors"] = $"/api/v1/documents/{documentId}/errors?historic={historic}",
            ["attachments"] = $"/api/v1/documents/{documentId}/attachments?historic={historic}",
            ["flags"] = $"/api/v1/documents/{documentId}/flags?historic={historic}",
            ["messages"] = $"/api/v1/documents/{documentId}/messages?historic={historic}",
        };

        return new DocumentDetailDto(
            doc.DocumentId, doc.DocumentType, doc.ProcessType,
            doc.CreationDate, doc.UpdateDate,
            doc.Owner, doc.OwnerSearchCode, doc.Receiver, doc.ReceiverSearchCode,
            doc.StatusId, doc.StatusCode, doc.StatusDescription, doc.Semaphore,
            doc.ConfigVersion, doc.ProcessVersion, doc.Historic,
            tags, tagDefs, actions, links);
    }

    public async Task<AttachmentListResponse?> GetAttachmentsAsync(string documentId, int historic = 0, CancellationToken ct = default)
    {
        var exists = await _db.Set<Document>().AnyAsync(d => d.DocumentId == documentId, ct);
        if (!exists) return null;

        var attachments = await _db.Set<DocumentAttachment>().AsNoTracking()
            .Where(a => a.DocumentId == documentId)
            .OrderByDescending(a => a.CreationDate)
            .ToListAsync(ct);

        var items = attachments.Select(a => new AttachmentDto(
            a.AttachmentId, a.AttachmentName, a.AttachmentName,
            a.Type, a.CreationDate, a.ContentType,
            a.Type == "Text",
            $"/api/v1/documents/{documentId}/attachments/{a.AttachmentName}/download?historic={historic}",
            a.Type == "Text" ? $"/api/v1/documents/{documentId}/attachments/{a.AttachmentName}/content?historic={historic}" : null
        )).ToList();

        return new AttachmentListResponse(documentId, items);
    }

    public async Task<AttachmentContentResponse?> GetAttachmentContentAsync(string documentId, string attachmentName, int historic = 0, CancellationToken ct = default)
    {
        var att = await _db.Set<DocumentAttachment>().AsNoTracking()
            .FirstOrDefaultAsync(a => a.DocumentId == documentId && a.AttachmentName == attachmentName, ct);
        if (att == null) return null;
        if (att.Type != "Text") return null;
        return new AttachmentContentResponse(att.AttachmentName, att.ContentType, att.Content ?? "");
    }

    public async Task<byte[]?> DownloadAttachmentAsync(string documentId, string attachmentName, int historic = 0, CancellationToken ct = default)
    {
        var att = await _db.Set<DocumentAttachment>().AsNoTracking()
            .FirstOrDefaultAsync(a => a.DocumentId == documentId && a.AttachmentName == attachmentName, ct);
        if (att == null) return null;
        return System.Text.Encoding.UTF8.GetBytes(att.Content ?? "");
    }

    public async Task<HistoryResponse?> GetHistoryAsync(string documentId, int historic = 0, CancellationToken ct = default)
    {
        var exists = await _db.Set<Document>().AnyAsync(d => d.DocumentId == documentId, ct);
        if (!exists) return null;
        var items = await _db.Set<DocumentHistoryFlow>().AsNoTracking()
            .Where(h => h.DocumentId == documentId).OrderByDescending(h => h.CreationDate).ToListAsync(ct);
        return new HistoryResponse(documentId, items.Select(h =>
            new HistoryFlowDto(h.HistoryId, h.CreationDate, h.ProcessName, h.FlowId, h.StatusId, h.StatusCode, h.StatusDescription, h.StatusDetail, h.QueueName, h.Semaphore, h.LoteId)).ToList());
    }

    public async Task<ErrorsResponse?> GetErrorsAsync(string documentId, int historic = 0, CancellationToken ct = default)
    {
        var exists = await _db.Set<Document>().AnyAsync(d => d.DocumentId == documentId, ct);
        if (!exists) return null;
        var items = await _db.Set<DocumentError>().AsNoTracking()
            .Where(e => e.DocumentId == documentId).OrderByDescending(e => e.CreationDate).ToListAsync(ct);
        return new ErrorsResponse(documentId, items.Select(e =>
            new DocumentErrorDto(e.ErrorId, e.CreationDate, e.Method, e.Code, e.Message)).ToList());
    }

    public async Task<AuditResponse?> GetAuditAsync(string documentId, int historic = 0, CancellationToken ct = default)
    {
        var exists = await _db.Set<Document>().AnyAsync(d => d.DocumentId == documentId, ct);
        if (!exists) return null;
        var items = await _db.Set<DocumentAction>().AsNoTracking()
            .Where(a => a.DocumentId == documentId).OrderByDescending(a => a.CreationDate).ToListAsync(ct);
        return new AuditResponse(documentId, items.Select(a =>
            new ActionAuditDto(a.ActionId, a.CreationDate, a.Username, a.ActionDescription)).ToList());
    }

    public async Task<FlagsResponse?> GetFlagsAsync(string documentId, int historic = 0, CancellationToken ct = default)
    {
        var exists = await _db.Set<Document>().AnyAsync(d => d.DocumentId == documentId, ct);
        if (!exists) return null;
        var items = await _db.Set<DocumentFlag>().AsNoTracking()
            .Where(f => f.DocumentId == documentId).ToListAsync(ct);
        return new FlagsResponse(documentId, items.Select(f => new FlagDto(f.FlagName, f.FlagValue)).ToList());
    }

    public async Task<MessagesResponse?> GetMessagesAsync(string documentId, int historic = 0, CancellationToken ct = default)
    {
        var exists = await _db.Set<Document>().AnyAsync(d => d.DocumentId == documentId, ct);
        if (!exists) return null;
        var items = await _db.Set<DocumentMessage>().AsNoTracking()
            .Where(m => m.DocumentId == documentId).OrderByDescending(m => m.CreationDate).ToListAsync(ct);
        return new MessagesResponse(documentId, items.Select(m =>
            new MessageDto(m.MessageId, m.CreationDate, m.MessageType, m.Code, m.Description, m.CustomReturn, m.Sent, m.Error, m.ProcessType, m.TagVarchar1, m.TagVarchar2, m.TagVarchar3)).ToList());
    }

    private static IQueryable<Document> ApplyFilter(IQueryable<Document> query, SearchFilterDto f)
    {
        if (f.Type.Contains("String") && !string.IsNullOrWhiteSpace(f.Value))
        {
            var val = f.Value;
            if (f.Name == "TagVarchar1") query = val.Contains('%') ? query.Where(d => d.TagVarchar1 != null && d.TagVarchar1.Contains(val.Replace("%", ""))) : query.Where(d => d.TagVarchar1 == val);
            else if (f.Name == "TagVarchar2") query = query.Where(d => d.TagVarchar2 == val);
            else if (f.Name == "TagVarchar3") query = query.Where(d => d.TagVarchar3 == val);
        }
        else if (f.Type.Contains("Integer") && f.Values is { Count: 2 })
        {
            if (int.TryParse(f.Values[0], out var min) && int.TryParse(f.Values[1], out var max))
            {
                if (f.Name == "TagInt1") query = query.Where(d => d.TagInt1 >= min && d.TagInt1 <= max);
                else if (f.Name == "TagInt2") query = query.Where(d => d.TagInt2 >= min && d.TagInt2 <= max);
            }
        }
        else if (f.Type.Contains("DateTime") && f.Values is { Count: 2 })
        {
            if (DateTime.TryParse(f.Values[0], out var from) && DateTime.TryParse(f.Values[1], out var to))
            {
                if (f.Name == "TagDate1") query = query.Where(d => d.TagDate1 >= from && d.TagDate1 <= to.Date.AddDays(1).AddTicks(-1));
            }
        }
        return query;
    }

    private static Dictionary<string, object?> BuildTags(Document d, List<FilterDefinitionDto> defs)
    {
        var tags = new Dictionary<string, object?>();
        foreach (var def in defs.Where(f => f.IsColumn))
        {
            tags[def.Name] = def.Name switch
            {
                "TagVarchar1" => d.TagVarchar1, "TagVarchar2" => d.TagVarchar2, "TagVarchar3" => d.TagVarchar3,
                "TagVarchar4" => d.TagVarchar4, "TagVarchar5" => d.TagVarchar5,
                "TagInt1" => d.TagInt1, "TagInt2" => d.TagInt2, "TagInt3" => d.TagInt3,
                "TagInt5" => d.TagInt5,
                "TagFloat1" => d.TagFloat1,
                "TagDate1" => d.TagDate1,
                _ => null
            };
        }
        return tags;
    }
}
